src/spatial/spatial_massaction_jump.jl

```diff
@@ -0,0 +1,65 @@
+        reactant_stoch = nocopy ? reactant_stoch : copy(reactant_stoch)
+        for i in eachindex(reactant_stoch)
+            if useiszero && (length(reactant_stoch[i]) == 1) && iszero(reactant_stoch[i][1][1])
+            reactant_stoch[i] = typeof(reactant_stoch[i])()
+            end
+        end
+        if scale_rates && !isempty(uniform_rates)
+            scalerates!(uniform_rates, reactant_stoch)
+        end
+        if scale_rates && !isempty(spatial_rates)
+            scalerates!(spatial_rates, reactant_stoch[length(uniform_rates)+1:end])
+        end
+        new(uniform_rates, spatial_rates, reactant_stoch, net_stoch, param_mapper)
+    end
+
+end
+SpatialMassActionJump(urates::Vector{F}, srates::Matrix{F}, rs::S, ns::U, pmapper::V; scale_rates = true, useiszero = true, nocopy=false) where {F,S,U,V} = SpatialMassActionJump{F,S,U,V}(urates, srates, rs, ns, pmapper, scale_rates, useiszero, nocopy)
+SpatialMassActionJump(urates::Vector{F}, srates::Matrix{F}, rs, ns; scale_rates = true, useiszero = true, nocopy=false) where {F} = SpatialMassActionJump(urates, srates, rs, ns, nothing; scale_rates=scale_rates, useiszero=useiszero, nocopy=nocopy)
+
+SpatialMassActionJump(srates::Matrix{F}, rs, ns, pmapper; scale_rates = true, useiszero = true, nocopy=false) where {F} = SpatialMassActionJump(zeros(F, 0), srates, rs, ns, pmapper; scale_rates = scale_rates, useiszero = useiszero, nocopy=nocopy)
+SpatialMassActionJump(srates::Matrix{F}, rs, ns; scale_rates = true, useiszero = true, nocopy=false) where {F} = SpatialMassActionJump(zeros(F, 0), srates, rs, ns, nothing; scale_rates = scale_rates, useiszero = useiszero, nocopy=nocopy)
+
+SpatialMassActionJump(urates::Vector{F}, rs, ns, pmapper; scale_rates = true, useiszero = true, nocopy=false) where {F} = SpatialMassActionJump(urates, zeros(F, 0, 0), rs, ns, pmapper; scale_rates = scale_rates, useiszero = useiszero, nocopy=nocopy)
+SpatialMassActionJump(urates::Vector{F}, rs, ns; scale_rates = true, useiszero = true, nocopy=false) where {F} = SpatialMassActionJump(urates, zeros(F, 0, 0), rs, ns, nothing; scale_rates = scale_rates, useiszero = useiszero, nocopy=nocopy)
+
+SpatialMassActionJump(ma_jumps::MassActionJump{T,S,U,V}; scale_rates = true, useiszero = true, nocopy=false) where {T,S,U,V} = SpatialMassActionJump(ma_jumps.scaled_rates, ma_jumps.reactant_stoch, ma_jumps.net_stoch, ma_jumps.param_mapper; scale_rates = scale_rates, useiszero = useiszero, nocopy=nocopy)
+
+##############################################
+
+get_num_majumps(spatial_majump::SpatialMassActionJump) = length(spatial_majump.uniform_rates) + size(spatial_majump.spatial_rates, 1)
+using_params(spatial_majump::SpatialMassActionJump) = false
+
+function rate_at_site(rx, site, spatial_majump::SpatialMassActionJump)
+    num_unif_rxs = length(spatial_majump.uniform_rates)
+    rx <= num_unif_rxs ? spatial_majump.uniform_rates[rx] : spatial_majump.spatial_rates[rx-num_unif_rxs, site]
```

Comment:
Weren't we going to allow one of `uniform_rates` or `spatial_rates` to be `nothing`, so dispatch could avoid having to branch here if just one of them is being used?